### ISSUE TITLE
static assertion failed: HRectBound can only be used with the LMetric<> metric type.

### ISSUE THREAD

**Issue by dongli** [mlpack/mlpack#525]
The documentation (http://www.mlpack.org/doxygen.php?doc=rstutorial.html) says
```
template<
  typename MetricType = mlpack::metric::EuclideanDistance,
  typename TreeType = mlpack::tree::BinarySpaceTree<bound::HRectBound<2>,
                                                           ^^^^^^^^^^^^^
                                                    tree::EmptyStatistic>
>
class RangeSearch;
```
where `HRectBound<2>` is not valid anymore, `g++` (5.2.0) complains:
```
error: type/value mismatch at argument 1 in template parameter list for 'template<class MetricType> class mlpack::bound::HRectBound'
     tree::BinarySpaceTree<bound::HRectBound<2>, range::RangeSearchStat> datasetTree(dataset);
                                              ^
/Users/dongli/Dropbox/Works/CxxExamples/src/UseLibraries/MLPACK/mlpack_exam03.cpp:56:46: note:   expected a type, got '2'
/Users/dongli/Dropbox/Works/CxxExamples/src/UseLibraries/MLPACK/mlpack_exam03.cpp:56:71: error: template argument 1 is invalid
     tree::BinarySpaceTree<bound::HRectBound<2>, range::RangeSearchStat> datasetTree(dataset);

```
after read the code `mlpack/core/tree/hrectbound.hpp`, I found `HRectBound` needs `MetricType` template argument. So I changed `bound::HRectBound<2>` to `bound::HRectBound<metric::LMetric<2, true>>`, then I got the new error:
```
error: static assertion failed: HRectBound can only be used with the LMetric<> metric type.
   static_assert(meta::IsLMetric<MetricType>::Value == true,
```
but the MetricType is a LMetric.

**Comment by rcurtin** [comment]
It looks like you're using an old version of the tutorial... is there a link somewhere on mlpack.org that went to that page?  (There shouldn't be, so if there is, I'll fix it.)

Take a look at the most recent version of the tutorial instead:
http://www.mlpack.org/docs/mlpack-2.0.1/doxygen.php?doc=rstutorial.html

The `RangeSearch` class should be used like this (now it uses template template parameters):

```
RangeSearch<LMetric<2, true>, arma::mat, tree::KDTree> rs;
```

I'm not sure why the `static_assert` is triggering.  What version of mlpack are you using?

**Comment by dongli** [comment]
I am using `2.0.1`. I followed the newest tutorial you provided, where it contains:
```
// Construct a RangeSearch object with ball bounds.
RangeSearch<
    metric::EuclideanDistance,
    tree::BinarySpaceTree<bound::BallBound<2>,
                          EmptyStatistic>
> rangeSearch(dataset);
```
but the compiler complains:
```
error: 
      template argument for template type parameter must be a type
    tree::BinarySpaceTree<bound::BallBound<2>, EmptyStatistic> datasetTr...
```

**Comment by dongli** [comment]
When I changed `bound::BallBound<2>` to `bound::BallBound<metric::LMetric<2>>`, that error is gone, but there is still error with `bound::HRectBound<metric::LMetric<2>>`. I think it is caused by the static assert in `hrectbound.hpp`:
```
 29 //! Utility struct where Value is true if and only if the argument is of type
 30 //! LMetric.
 31 template<typename MetricType>
 32 struct IsLMetric
 33 {
 34   static const bool Value = false;
 35 };
 36 
 37 //! Specialization for IsLMetric when the argument is of type LMetric.
 38 template<int Power, bool TakeRoot>
 39 struct IsLMetric<metric::LMetric<Power, TakeRoot>>
 40 {
 41   static const bool Value = true;
 42 };
 43 
 44 } // namespace util
```

**Comment by dongli** [comment]
I think the correct usage of `BinarySpaceTree` is `tree::BinarySpaceTree<metric::EuclideanDistance>` where the default bound template class is `bound::HRectBound`. So is the tutorial wrong, which says `tree::BinarySpaceTree<bound::BallBound<2>`?

**Comment by rcurtin** [comment]
Yes, the tutorial was wrong; it was fixed some time ago in 79fd639 (but this was apparently after 2.0.1, so these are only updated in the git master branch documentation: http://mlpack.org/docs/mlpack-git/doxygen.php?doc=rstutorial.html ; side note: I need to rebuild those with an older version of Doxygen so the CSS styling is more correct...).

I also noticed that the mlpack website had some incorrect links on help.html to old versions of documentation; this is fixed in https://github.com/mlpack/mlpack.org/commit/abf55c206d9702811794e1f6f71db46f540d6339 .

So the problem here wasn't the static_assert but instead that the documentation was out of date, and the tree classes are used a bit differently now.  Can you try with the new tutorial code and see if that fixes things?

**Comment by rcurtin** [comment]
Closing for inactivity... if you'd like to reopen this, we can, but I think the issue is solved.
